File: QuickSort/Python/QuickSort_Performance.py
# QuickSort_Performance.py
"""
Performance benchmark harness for Python sorting algorithms.

Assumes each algorithm is a function:
    def quicksort(a: list[int]) -> None
"""

from __future__ import annotations
from typing import Callable, Dict, List, Tuple
import random
import time
import statistics
import csv

# Import Algorithms

from QuickSort_ChatGPT import quicksort as quicksort_chatgpt
from QuickSort_Claude import quicksort as quicksort_claude
from QuickSort_DeepSeek import quicksort as quicksort_deepseek
from QuickSort_Gemini import quicksort as quicksort_gemini

SortFunc = Callable[[List[int]], None]

ALGORITHMS: Dict[str, SortFunc] = {
    "chatgpt" : quicksort_chatgpt,
    "claude" : quicksort_claude,
    "deepseek" : quicksort_deepseek,
    "gemini" : quicksort_gemini,
}

# Test Case Generators

def gen_random(n: int) -> List[int]:
    return [random.randint(-10**6, 10**6) for _ in range(n)]

def gen_sorted(n: int) -> List[int]:
    arr = gen_random(n)
    arr.sort()
    return arr

def gen_reversed(n: int) -> List[int]:
    arr = gen_random(n)
    arr.sort(reverse=True)
    return arr

def gen_nearly_sorted(n: int) -> List[int]:
    """Start sorted, then do a few random swaps."""
    arr = gen_sorted(n)
    if n <= 1:
        return arr
    num_swaps = max(1, n // 100)
    for _ in range(num_swaps):
        i = random.randrange(n)
        j = random.randrange(n)
        arr[i], arr[j] = arr[j], arr[i]
    return arr

def gen_few_values(n: int) -> List[int]:
    """Many duplicates: values from a small range."""
    return [random.randint(0, 9) for _ in range(n)]

DISTRIBUTIONS: Dict[str, Callable[[int], List[int]]] = {
    "random": gen_random,
    "sorted": gen_sorted,
    "reversed": gen_reversed,
    "nearly_sorted": gen_nearly_sorted,
    "few_values": gen_few_values,
}

SIZES = [50, 100, 200, 500]
RUNS_PER_COMBO = 10  # how many timing runs per (algo, dist, size)

# Benchmark logic

def precompute_inputs() -> Dict[Tuple[str, int, int], List[int]]:
    """
    Precompute base arrays so every algorithm sees identical inputs.

    Key is (dist_name, size, run_index) -> list[int]
    """
    inputs: Dict[Tuple[str, int, int], List[int]] = {}
    for dist_name, generator in DISTRIBUTIONS.items():
        for n in SIZES:
            for run_idx in range(RUNS_PER_COMBO):
                inputs[(dist_name, n, run_idx)] = generator(n)
    return inputs

def time_one_run(sort_func: SortFunc, arr: List[int]) -> float:
    """Time a single call to sort_func on a copy of arr."""
    a = list(arr)  
    t0 = time.perf_counter()
    sort_func(a)
    t1 = time.perf_counter()
    return t1 - t0

def benchmark() -> List[Dict[str, object]]:

    random.seed(1)  # for reproducibility
    base_inputs = precompute_inputs()

    results: List[Dict[str, object]] = []

    for algo_name, sort_func in ALGORITHMS.items():
        print(f"Benchmarking algorithm: {algo_name}")

        for dist_name in DISTRIBUTIONS.keys():
            for n in SIZES:
                durations: List[float] = []

                # multiple runs
                for run_idx in range(RUNS_PER_COMBO):
                    arr = base_inputs[(dist_name, n, run_idx)]
                    dt = time_one_run(sort_func, arr)
                    durations.append(dt)

                durations.sort()
                median = statistics.median(durations)
                mean = statistics.mean(durations)
                min_t = durations[0]
                max_t = durations[-1]

                print(
                    f"{dist_name:13s} | n={n:4d} | "
                    f"median={median:.6f}s | mean={mean:.6f}s | "
                    f"min={min_t:.6f}s | max={max_t:.6f}s"
                )

                results.append(
                    {
                        "algorithm": algo_name,
                        "distribution": dist_name,
                        "n": n,
                        "runs": RUNS_PER_COMBO,
                        "median_sec": median,
                        "mean_sec": mean,
                        "min_sec": min_t,
                        "max_sec": max_t,
                    }
                )

        print()

    print("All benchmarks completed.")
    return results

def write_csv(results: List[Dict[str, object]], filename: str = "python_bench_results.csv") -> None:
    if not results:
        print("No results to write.")
        return

    fieldnames = list(results[0].keys())
    with open(filename, "w", newline="") as f:
        writer = csv.DictWriter(f, fieldnames=fieldnames)
        writer.writeheader()
        for row in results:
            writer.writerow(row)

    print(f"Results written to {filename}")

if __name__ == "__main__":
    res = benchmark()
    write_csv(res)

File: QuickSort/Rust/src/bin/quicksort_performance.rs
// src/bin/quicksort_performance.rs
//
// Performance benchmark harness for Rust sorting algorithms.
//
// Run with:
//    cargo run --bin quicksort_performance --release
//
// and writes results to "rust_bench_results.csv" in the project root.

use std::fs::File;
use rand::rngs::StdRng;
use rand::{Rng, SeedableRng};
use std::io::{BufWriter, Write};
use std::time::Instant;

type SortFn = fn(&mut [i32]);

// Import Algorithms

mod quicksort_chatgpt;
mod quicksort_claude;
mod quicksort_deepseek;
mod quicksort_gemini;

use quicksort_chatgpt::quicksort as quicksort_chatgpt_fn;
use quicksort_claude::quicksort as quicksort_claude_fn;
use quicksort_deepseek::quicksort as quicksort_deepseek_fn;
use quicksort_gemini::quicksort as quicksort_gemini_fn;

// Test Case Generators

fn gen_random(n: usize, rng: &mut StdRng) -> Vec<i32> {
    (0..n)
        .map(|_| rng.gen_range(-1_000_000..=1_000_000))
        .collect()
}

fn gen_sorted(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_random(n, rng);
    v.sort();
    v
}

fn gen_reversed(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_random(n, rng);
    v.sort_by(|a, b| b.cmp(a));
    v
}

fn gen_nearly_sorted(n: usize, rng: &mut StdRng) -> Vec<i32> {
    let mut v = gen_sorted(n, rng);
    if n <= 1 {
        return v;
    }
    let num_swaps = (n / 100).max(1); // ~1%
    for _ in 0..num_swaps {
        let i = rng.gen_range(0..n);
        let j = rng.gen_range(0..n);
        v.swap(i, j);
    }
    v
}

fn gen_few_values(n: usize, rng: &mut StdRng) -> Vec<i32> {
    (0..n).map(|_| rng.gen_range(0..=9)).collect()
}

const DIST_NAMES: &[&str] = &["random", "sorted", "reversed", "nearly_sorted", "few_values"];
const SIZES: &[usize] = &[500, 1_000, 5_000, 10_000];
const RUNS_PER_COMBO: usize = 10;

// Precomputed inputs so each algorithm sees identical data.
struct InputSet {
    dist_name: &'static str,
    n: usize,
    runs: Vec<Vec<i32>>,
}

fn precompute_inputs() -> Vec<InputSet> {
    println!("Precomputing input arrays...");
    let mut rng = StdRng::seed_from_u64(1);
    let mut sets = Vec::new();

    for &dist in DIST_NAMES {
        for &n in SIZES {
            let mut runs = Vec::with_capacity(RUNS_PER_COMBO);
            for _ in 0..RUNS_PER_COMBO {
                let arr = match dist {
                    "random" => gen_random(n, &mut rng),
                    "sorted" => gen_sorted(n, &mut rng),
                    "reversed" => gen_reversed(n, &mut rng),
                    "nearly_sorted" => gen_nearly_sorted(n, &mut rng),
                    "few_values" => gen_few_values(n, &mut rng),
                    _ => panic!("unknown distribution {dist}"),
                };
                runs.push(arr);
            }
            sets.push(InputSet { dist_name: dist, n, runs });
        }
    }

    println!("Done.\n");
    sets
}

// Benchmark logic

#[derive(Debug)]
struct BenchRow {
    algorithm: String,
    distribution: String,
    n: usize,
    runs: usize,
    median_sec: f64,
    mean_sec: f64,
    min_sec: f64,
    max_sec: f64,
}

fn time_one_run(sort_fn: SortFn, arr: &[i32]) -> f64 {
    let mut v = arr.to_vec();
    let start = Instant::now();
    sort_fn(&mut v);
    let dt = start.elapsed();
    dt.as_secs_f64()
}

fn benchmark() -> Vec<BenchRow> {
    let algos = vec![
        ("Chatgpt", quicksort_chatgpt_fn as SortFn),
        ("Claude", quicksort_claude_fn as SortFn),
        ("Deepseek", quicksort_deepseek_fn as SortFn),
        ("Gemini", quicksort_gemini_fn as SortFn),
    ];

    if algos.is_empty() {
        eprintln!("No algorithms registered.");
        return Vec::new();
    }

    let inputs = precompute_inputs();
    let mut rows = Vec::new();

    for (name, sort_fn) in algos {
        println!("Benchmarking algorithm: {name}");

        for set in &inputs {
            let mut durations = Vec::with_capacity(set.runs.len());

            for arr in &set.runs {
                let dt = time_one_run(sort_fn, arr);
                durations.push(dt);
            }

            durations.sort_by(|a, b| a.partial_cmp(b).unwrap());
            let min = durations[0];
            let max = *durations.last().unwrap();
            let median = if durations.len() % 2 == 0 {
                let mid = durations.len() / 2;
                (durations[mid - 1] + durations[mid]) / 2.0
            } else {
                durations[durations.len() / 2]
            };
            let sum: f64 = durations.iter().copied().sum();
            let mean = sum / durations.len() as f64;

            println!(
                "{:<12} | {:<13} | n={:<7} | median={:.6}s | mean={:.6}s | min={:.6}s | max={:.6}s",
                name, set.dist_name, set.n, median, mean, min, max
            );

            rows.push(BenchRow {
                algorithm: name.to_string(),
                distribution: set.dist_name.to_string(),
                n: set.n,
                runs: durations.len(),
                median_sec: median,
                mean_sec: mean,
                min_sec: min,
                max_sec: max,
            });
        }

        println!();
    }

    println!("All benchmarks completed.");
    rows
}

fn write_csv(rows: &[BenchRow], filename: &str) {
    if rows.is_empty() {
        eprintln!("No rows to write.");
        return;
    }

    let file = File::create(filename).expect("failed to create CSV file");
    let mut w = BufWriter::new(file);

    writeln!(
        w,
        "algorithm,distribution,n,runs,median_sec,mean_sec,min_sec,max_sec"
    )
    .unwrap();

    for r in rows {
        writeln!(
            w,
            "{},{},{},{},{:.9},{:.9},{:.9},{:.9}",
            r.algorithm,
            r.distribution,
            r.n,
            r.runs,
            r.median_sec,
            r.mean_sec,
            r.min_sec,
            r.max_sec
        )
        .unwrap();
    }

    println!("Results written to {filename}");
}

fn main() {
    let rows = benchmark();
    write_csv(&rows, "rust_bench_results.csv");
}
